flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/TumblingProcessingTimeWindows.java

```diff
@@ -46,28 +46,44 @@
 
 	private final long size;
 
-	private final long offset;
+	private final long globalOffset;
 
-	private TumblingProcessingTimeWindows(long size, long offset) {
+	private Long staggerOffset = null;
+
+	private final WindowStagger windowStagger;
+
+	private TumblingProcessingTimeWindows(long size, long offset, WindowStagger windowStagger) {
 		if (Math.abs(offset) >= size) {
 			throw new IllegalArgumentException("TumblingProcessingTimeWindows parameters must satisfy abs(offset) < size");
 		}
 
 		this.size = size;
-		this.offset = offset;
+		this.globalOffset = offset;
+		this.windowStagger = windowStagger;
 	}
 
 	@Override
 	public Collection<TimeWindow> assignWindows(Object element, long timestamp, WindowAssignerContext context) {
 		final long now = context.getCurrentProcessingTime();
-		long start = TimeWindow.getWindowStartWithOffset(now, offset, size);
+		if (staggerOffset == null) {
+			staggerOffset = windowStagger.getStaggerOffset(context.getCurrentProcessingTime(), size);
```

Comment:
I think my comment from your other PR still hold: Does WindowStagger.NATURAL make sense if we keep the stagger offset fixed per
WindowAssigner instance? I don't know if this was the intention but that's what
the code does.

This will set a fixed offset when the first event is received, Flink will not create a new `WindowAssigner` instance for new panes, so those new panes will use the same offset that was set for the first pane that we ever saw. That's why I'm saying that `NATURAL` doesn't really do anything here. (Or I'm misunderstanding sth.)

Reply:
Yes, your interpretation is correct, the intention was to align the panes with respect to the first event evert ingested.

Events arrive at different times among partitions, the differences were usually caused by the design of partition keys, network delay, etc, which led to an implicit staggering (normal distribution in our case).

This is useful because compared to the ALIGNED and RANDOM, this one staggers the pane but still preserves some useful alignments (for example, in our geospatial applications, windowing on events partitioned by cities still trigger at the same time if they're under same time zone ).

Therefore, we think this would be a useful option for staggering.

Reply:
Ok, now I understand. Thanks!